### PR TITLE
Add NMSv3 GPU op

### PR DESCRIPTION
**Pull request by samikama**
This PR adds GPU version of NMSv3 Op and extends NMS kernel to support legacy networks as well as modern networks, addressing the issue mentioned in original NMS GPU op commit.

### CODE REVIEW

**Comment by samikama** [comment]
@aaroey, can you take a look. This is pretty much the same of NMSv2 except that the NMS op is limited to boxes with scores greater than the threshold.

**Comment by fferroni** [comment]
@samikama  hi, do you have an expectation of the speed-up from CUDA? I've been trying on some synthetic data of 100,000 boxes, and get a CPU time that is much better than your implementation...?

**Comment by aaroey** [comment]
@samikama could you address the comment by @fferroni above and provide a simple repro of the performance benefits?

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by samikama** [comment]
@fferroni Could you post your test case? In my tests it was on par with CPU implementation. Nevertheless, I did improve the performance with a new commit. Here are some numbers from my tests
```
python TestNMSv3.py --num_iter=1000 -num_boxes=4096 --max_output=2000
GPUv3     times=  1.196 +-   0.040 msec 
GPUv2     times=  1.170 +-   0.468 msec 
CPU         times= 77.669 +-  17.420 msec
HCPU      times= 75.239 +-  17.725 msec 
```
Test randomly generates 4096 boxes and scores that are fed into NMSv2 and NMSv3 ops on GPU and NMSV3 op on CPU, max_output_size is set to 2000 in all ops, data is placed in variables so no host to device copy needed. CPU is fed with data on GPU, HCPU is fed with data on CPU. All arguments of the ops are identical whenever possible. 1000 tests are run and average and stddev is reported. There can be further improvements that will show up when op is executed on graph by changing op to an async op but considering the improvement, I don't think it is necessary.

**Comment by samikama** [comment]
@aaroey do you have any more concerns?

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 1]
```diff
@@ -80,12 +80,22 @@ __device__ EIGEN_STRONG_INLINE void Swap(T& a, T& b) {
   b = c;
 }
 
+template <bool T>
+__device__ float legacy_offset(float);
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: As commented on the other PR, please remove these legacy offset funcs?

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 2]
```diff
@@ -118,6 +128,47 @@ __device__ EIGEN_STRONG_INLINE void Flipped<true>(Box& box) {
   if (box.x1 > box.x2) Swap(box.x1, box.x2);
   if (box.y1 > box.y2) Swap(box.y1, box.y2);
 }
+template <typename T>
+__device__ EIGEN_STRONG_INLINE bool CheckBit(T* bit_mask, int bit) {
+  constexpr int SHIFTLEN = NumBits(8 * sizeof(T)) - 1;
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: nit: kShiftLen, kRemainderMask

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 3]
```diff
@@ -487,9 +554,241 @@ class NonMaxSuppressionV2GPUOp : public OpKernel {
...
+  TF_RETURN_IF_CUDA_ERROR(cudaEventSynchronize(copy_done));
+  return Status::OK();
+}
+
+class NonMaxSuppressionV3GPUOp : public OpKernel {
+ public:
+  explicit NonMaxSuppressionV3GPUOp(OpKernelConstruction* context)
+      : OpKernel(context) {}
+
+  void Compute(OpKernelContext* context) override {
```
Comment (aaroey) [inline tensorflow/core/kernels/non_max_suppression_op.cu.cc]: This method is almost the same as NonMaxSuppressionV2GPUOp::Compute() except the additional score input handling logic. Please do something like: add a NonMaxSuppressionGPUOpBase class and encapsulate the common code as helper member functions and reuse them in both ops.

**Review by aaroey — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/non_max_suppression_op.cu.cc [hunk 4]
```diff
@@ -16,17 +16,17 @@ limitations under the License.
...
-#include "third_party/cub/device/device_segmented_radix_sort.cuh"
-#include "third_party/cub/device/device_select.cuh"
 #include "tensorflow/core/framework/numeric_types.h"
 #include "tensorflow/core/framework/op_kernel.h"
 #include "tensorflow/core/framework/tensor_types.h"
 #include "tensorflow/core/kernels/non_max_suppression_op.h"
 #include "tensorflow/core/util/gpu_kernel_helper.h"
 #include "tensorflow/core/util/gpu_launch_config.h"
 #include "tensorflow/stream_executor/stream_executor.h"
+#include "third_party/cub/device/device_radix_sort.cuh"
```
[the remark this answers is not in the document]
Comment (samikama) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: Hi @aaroey These were changed by clang-format, If I revert them clang-format check will complain.

[the remark this answers is not in the document]
Comment (aaroey) [reply tensorflow/core/kernels/non_max_suppression_op.cu.cc]: I see, sorry I didn't know that. Feel free to skip.

**Review by aaroey — DISMISSED** [review]
Thanks for the fix!

**Review by aaroey — APPROVED** [review]

**Comment by aaroey** [comment]
This PR breaks an internal test and was partially rolled back in https://github.com/tensorflow/tensorflow/commit/009dcac2213e737f22db69e84615b66f77bd6ddb, and the problem is tracked in #32446.

**Pull request merged**